File: apps/pwa/app/matrixNativeProjection.test.ts
import { describe, expect, it } from "vitest";
import { MatrixNativeProjection } from "./matrixNativeProjection";

describe("MatrixNativeProjection Room State", () => {
  it("builds inventory from Gateway and per-session current state", async () => {
    const projection = new MatrixNativeProjection();
    const entity = sessionState("s1", 2, {
      title: "Investigate sync",
      status: "running",
    });
    await projection.applyRoomState(gatewayState(2));
    const state = await projection.applyRoomState(entity);
    expect(state?.sessions).toMatchObject([{
      id: "s1",
      title: "Investigate sync",
      status: "running",
    }]);
  });

  it("accepts current Room State whether Gateway or session arrives first", async () => {
    const gatewayFirst = new MatrixNativeProjection();
    await expect(gatewayFirst.applyRoomState(gatewayState())).resolves.toMatchObject({
      sessions: [],
    });
    await expect(gatewayFirst.applyRoomState(sessionState("s1", 2))).resolves.toMatchObject({
      sessions: [{ id: "s1" }],
    });

    const sessionFirst = new MatrixNativeProjection();
    await expect(sessionFirst.applyRoomState(sessionState("s1", 2))).resolves.toBeNull();
    await expect(sessionFirst.applyRoomState(gatewayState())).resolves.toMatchObject({
      sessions: [{ id: "s1" }],
    });
  });

  it("publishes one complete Room State batch and rolls back invalid batches", async () => {
    const projection = new MatrixNativeProjection();
    await expect(projection.applyRoomStateBatch([
      gatewayState(2),
      sessionState("s1", 2),
      sessionState("s2", 3),
    ])).resolves.toMatchObject({
      sessions: [{ id: "s2" }, { id: "s1" }],
    });

    const before = projection.snapshot();
    await expect(projection.applyRoomStateBatch([
      gatewayState(5),
      tombstone("s1", 4),
      { ...sessionState("broken", 5), state: "invalid" },
    ])).rejects.toThrow();
    expect(projection.snapshot()).toEqual(before);
  });

  it("replaces one entity without scanning or replaying timeline history", async () => {
    const projection = new MatrixNativeProjection();
    const before = sessionState("s1", 2, { title: "Before" });
    const after = sessionState("s1", 3, {
      title: "After",
      status: "running",
    });
    await projection.applyRoomState(gatewayState(2));
    await projection.applyRoomState(before);
    const state = await projection.applyRoomState(after);
    expect(state?.sessions).toMatchObject([{ title: "After", status: "running" }]);
  });

  it("replaces missing directory entries but preserves newer live state", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomStateBatch([
      gatewayState(2),
      sessionState("kept", 2),
      sessionState("removed", 2),
    ]);
    await projection.applyRoomState(sessionState("arrived-live", 4));
    const committed = {
      ...gatewayState(5),
      session_directory: { ...gatewayState(5).session_directory, state_version: 3 },
    };
    const snapshot = await projection.applyRoomStateBatch([
      committed,
      sessionState("kept", 3),
    ]);
    expect(snapshot?.sessions.map((session) => session.id).sort()).toEqual([
      "arrived-live",
      "kept",
    ]);
  });

  it("requires a new authoritative batch when the Gateway directory commit changes", async () => {
    const projection = new MatrixNativeProjection();
    const current = gatewayState(2);
    await projection.applyRoomStateBatch([current, sessionState("removed", 2)]);
    expect(projection.requiresAuthoritativeDirectoryRefresh({
      ...current,
      updated_at: 42,
    })).toBe(false);
    const committed = {
      ...gatewayState(5),
      session_directory: {
        ...gatewayState(5).session_directory,
        state_version: 5,
        digest: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
      },
    };
    expect(projection.requiresAuthoritativeDirectoryRefresh(committed)).toBe(true);
    expect(projection.requiresCommandScopeRefresh(committed)).toBe(false);
    await projection.applyRoomStateBatch([committed]);
    expect(projection.snapshot()?.sessions).toEqual([]);
  });

  it("does not let an older directory fetch overwrite a newer live commit", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomStateBatch([
      gatewayState(2),
      sessionState("removed", 2),
    ]);
    await projection.applyRoomState(gatewayState(5));
    await expect(projection.applyRoomStateBatch([
      gatewayState(3),
      sessionState("removed", 3),
    ])).rejects.toThrow("advanced while an older snapshot was loading");
    expect(projection.requiresAuthoritativeDirectoryRefresh(gatewayState(5))).toBe(false);
    await projection.applyRoomStateBatch([gatewayState(5)]);
    expect(projection.snapshot()?.sessions).toEqual([]);
  });

  it("updates one session without waiting for an unrelated global commit", async () => {
    const projection = new MatrixNativeProjection();
    const before = sessionState("s1", 2, { title: "Before" });
    const after = sessionState("s1", 3, { title: "After" });
    await projection.applyRoomState(before);
    await projection.applyRoomState(gatewayState(2));
    await expect(projection.applyRoomState(after)).resolves.toMatchObject({
      sessions: [{ title: "After" }],
    });
  });

  it("keeps a tombstone authoritative over older session state", async () => {
    const projection = new MatrixNativeProjection();
    const current = sessionState("s1", 2);
    const deleted = tombstone("s1", 4);
    await projection.applyRoomState(gatewayState(2));
    await projection.applyRoomState(current);
    await projection.applyRoomState(deleted);
    await projection.applyRoomState(sessionState("s1", 3, { title: "Stale" }));
    expect(projection.snapshot()?.sessions).toEqual([]);
    expect(projection.sessionLifecycleState("s1")).toBe("deleted");
  });

  it("exposes the lifecycle value that actually won projection ordering", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomState(gatewayState(2));
    await projection.applyRoomState(sessionState("s1", 5));
    await projection.applyRoomState(tombstone("s1", 4));

    expect(projection.sessionLifecycleState("s1")).toBe("active");
    expect(projection.snapshot()?.sessions).toHaveLength(1);
  });

  it("does not let timeline roots create inventory entities", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomState(gatewayState(1));
    projection.applyTimeline({
      version: 2,
      kind: "session_root",
      ...revision(2),
      session_id: "timeline-only",
      title: "Must not appear",
      project: { id: "p1", name: "malink", cwd: "/repo" },
      created_at: 1,
      updated_at: 1,
      archived: false,
      status: "idle",
      provider: "codex",
      permission_mode: "default",
      extensions: [],
    });
    expect(projection.snapshot()?.sessions).toEqual([]);
  });

  it("advances command revision from a lightweight timeline event", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomState(gatewayState(1));
    const state = projection.applyTimeline({
      version: 2,
      kind: "gateway_revision",
      ...revision(4),
      gateway_id: "g1",
      conversation_id: "c1",
      updated_at: 4,
      source_command_id: "command-4",
    });
    expect(state?.revision).toBe(4);
  });

  it("projects a fresh Gateway heartbeat without advancing semantic state", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomState(gatewayState(1));
    const heartbeat = await projection.applyRoomState({
      ...gatewayState(1),
      updated_at: 42,
    });
    expect(heartbeat).toMatchObject({
      stateVersion: 1,
      revision: 1,
      updatedAt: 42,
    });
  });

  it("drops retired-epoch entities when Gateway generation advances", async () => {
    const projection = new MatrixNativeProjection();
    const old = sessionState("old", 2);
    await projection.applyRoomState(gatewayState(2));
    await projection.applyRoomState(old);
    await projection.applyRoomState({
      ...gatewayState(1),
      revision_epoch: "r2",
      revision_epoch_generation: 2,
      state_version: 1,
    });
    expect(projection.snapshot()?.sessions).toEqual([]);
  });

  it("allows live presentation status only for an existing Room State entity", async () => {
    const projection = new MatrixNativeProjection();
    await projection.applyRoomState(gatewayState(1));
    projection.applySessionStatus({
      kind: "status",
      session_id: "missing",
      state: "running",
    });
    expect(projection.snapshot()?.sessions).toEqual([]);
    const entity = sessionState("s1", 2);
    await projection.applyRoomState(entity);
    await projection.applyRoomState(gatewayState(2));
    expect(projection.applySessionStatus({
      kind: "status",
      session_id: "s1",
      state: "running",
      activity_phase: "working",
    })?.sessions).toMatchObject([{ status: "running", activityPhase: "working" }]);
  });
});

function gatewayState(stateVersion = 1) {
  return {
    version: 2,
    kind: "gateway_state",
    gateway_id: "g1",
    conversation_id: "c1",
    ...revision(1),
    state_version: stateVersion,
    active_device_count: 1,
    command_sequences: [
      { device_id: "device-1", sequence_epoch: "certificate-1", sequence: 0 },
    ],
    workspace: {
      project: { id: "p1", name: "malink", cwd: "/repo" },
      provider: "codex",
      permission_mode: "default",
    },
    capabilities: {
      models: [],
      permission_modes: [{ id: "default", name: "Default" }],
      can_create_session: true,
      can_select_session: false,
      can_archive_session: true,
      can_delete_session: true,
      session_extensions: [],
    },
    session_directory: {
      generation: stateVersion,
      state_version: stateVersion,
      slot: stateVersion % 3,
      page_count: 0,
      state_key_prefix: "malink.directory",
      digest: "RBNvo1WzZ4oRRq0W9-hknpT7T8If536DEMBg9hyq_4o",
    },
    updated_at: 1,
  } as const;
}

function sessionState(
  sessionId: string,
  stateVersion: number,
  overrides: { title?: string; status?: "idle" | "running" } = {},
) {
  return {
    version: 2,
    kind: "session_state",
    gateway_id: "g1",
    conversation_id: "c1",
    ...revision(stateVersion),
    state_version: stateVersion,
    session_id: sessionId,
    state: "active",
    session: {
      session_id: sessionId,
      thread_root_event_id: `$${sessionId}:example.org`,
      title: overrides.title ?? sessionId,
      updated_at: stateVersion,
      archived: false,
      status: overrides.status ?? "idle",
      project: { id: "p1", name: "malink", cwd: "/repo" },
      provider: "codex",
      extensions: [],
    },
    updated_at: stateVersion,
  } as const;
}

function tombstone(sessionId: string, stateVersion: number) {
  return {
    version: 2,
    kind: "session_state",
    gateway_id: "g1",
    conversation_id: "c1",
    ...revision(stateVersion),
    state_version: stateVersion,
    session_id: sessionId,
    state: "deleted",
    updated_at: stateVersion,
  } as const;
}

function revision(value: number) {
  return {
    revision: value,
    revision_epoch: "r1",
    revision_epoch_generation: 1,
  } as const;
}

File: clients/android/app/src/test/java/id/my/anciety/malink/client/MatrixNativeProjectionTest.kt
package id.my.anciety.malink.client

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Test

class MatrixNativeProjectionTest {
    @Test
    fun `builds inventory from independent Gateway and session Room State`() {
        val projection = MatrixNativeProjection()
        val entity = sessionState("session-1", 2)
        assertNull(projection.applyRoomState(entity))
        val state = projection.applyRoomState(gatewayState(2))!!
        assertEquals("session-1", state.sessions().single().getValue("id").jsonPrimitive.content)
    }

    @Test
    fun `Gateway first exposes an empty view then accepts session state independently`() {
        val projection = MatrixNativeProjection()
        assertEquals(0, projection.applyRoomState(gatewayState(1))!!.sessions().size)
        val state = projection.applyRoomState(sessionState("session-1", 2))!!
        assertEquals("session-1", state.sessions().single().getValue("id").jsonPrimitive.content)
    }

    @Test
    fun `complete Room State batch publishes atomically and invalid batch rolls back`() {
        val projection = MatrixNativeProjection()
        val snapshot = projection.applyRoomStateBatch(listOf(
            gatewayState(2),
            sessionState("session-1", 2),
            sessionState("session-2", 3),
        ))!!
        assertEquals(setOf("session-1", "session-2"), snapshot.sessions().map {
            it.getValue("id").jsonPrimitive.content
        }.toSet())

        val before = projection.snapshot()
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomStateBatch(listOf(
                gatewayState(5),
                tombstone("session-1", 4),
                JsonObject(sessionState("broken", 5) + mapOf(
                    "state" to JsonPrimitive("invalid"),
                )),
            ))
        }
        assertEquals(before, projection.snapshot())
    }

    @Test
    fun `updates one session without waiting for a global inventory commit`() {
        val projection = MatrixNativeProjection()
        val before = sessionState("session-1", 2, "Before")
        val after = sessionState("session-1", 3, "After")
        projection.applyRoomState(before)
        projection.applyRoomState(gatewayState(2))
        projection.applyRoomState(after)
        assertEquals(
            "After",
            projection.snapshot()!!.sessions().single()
                .getValue("title").jsonPrimitive.content,
        )
    }

    @Test
    fun `directory replacement removes missing entries but preserves newer live state`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomStateBatch(listOf(
            gatewayState(2),
            sessionState("kept", 2),
            sessionState("removed", 2),
        ))
        projection.applyRoomState(sessionState("arrived-live", 4))
        val committed = JsonObject(gatewayState(5) + mapOf(
            "session_directory" to buildJsonObject {
                put("generation", 5); put("state_version", 3); put("slot", 2)
                put("page_count", 1); put("state_key_prefix", "malink.directory")
                put("digest", "RBNvo1WzZ4oRRq0W9-hknpT7T8If536DEMBg9hyq_4o")
            },
        ))
        val snapshot = projection.applyRoomStateBatch(listOf(
            committed,
            sessionState("kept", 3),
        ))!!
        assertEquals(
            setOf("arrived-live", "kept"),
            snapshot.sessions().map { it.getValue("id").jsonPrimitive.content }.toSet(),
        )
    }

    @Test
    fun `Gateway directory commit requires a new authoritative batch`() {
        val projection = MatrixNativeProjection()
        val current = gatewayState(2)
        projection.applyRoomStateBatch(listOf(current, sessionState("removed", 2)))
        assertEquals(false, projection.requiresAuthoritativeDirectoryRefresh(
            JsonObject(current + mapOf("updated_at" to JsonPrimitive(42))),
        ))
        val committed = JsonObject(gatewayState(5) + mapOf(
            "session_directory" to buildJsonObject {
                put("generation", 5); put("state_version", 5); put("slot", 2)
                put("page_count", 0); put("state_key_prefix", "malink.directory")
                put("digest", "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
            },
        ))
        assertEquals(true, projection.requiresAuthoritativeDirectoryRefresh(committed))
        assertEquals(false, projection.requiresCommandScopeRefresh(committed))
        projection.applyRoomStateBatch(listOf(committed))
        assertEquals(0, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `older directory fetch cannot overwrite a newer live commit`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomStateBatch(listOf(
            gatewayState(2),
            sessionState("removed", 2),
        ))
        projection.applyRoomState(gatewayState(5))
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomStateBatch(listOf(
                gatewayState(3),
                sessionState("removed", 3),
            ))
        }
        assertEquals(false, projection.requiresAuthoritativeDirectoryRefresh(gatewayState(5)))
        projection.applyRoomStateBatch(listOf(gatewayState(5)))
        assertEquals(0, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `timeline history advances revision but never creates inventory`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomState(gatewayState(1))
        projection.applyTimeline(buildJsonObject {
            put("version", 2); put("kind", "session_root"); putRevision(4)
            put("session_id", "timeline-only")
        })
        assertEquals(0, projection.snapshot()!!.sessions().size)
        assertEquals("4", projection.snapshot()!!.getValue("revision").jsonPrimitive.content)
    }

    @Test
    fun `Gateway heartbeat refreshes public liveness without advancing semantic state`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomState(gatewayState(1))
        val heartbeat = JsonObject(gatewayState(1) + mapOf(
            "updated_at" to JsonPrimitive(42),
        ))
        val snapshot = projection.applyRoomState(heartbeat)!!
        assertEquals("1", snapshot.getValue("state_version").jsonPrimitive.content)
        assertEquals("1", snapshot.getValue("revision").jsonPrimitive.content)
        assertEquals("42", snapshot.getValue("updated_at").jsonPrimitive.content)
    }

    @Test
    fun `fresh authenticated session state also refreshes public Gateway liveness`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomState(gatewayState(1))
        val freshSession = JsonObject(sessionState("session-1", 2) + mapOf(
            "updated_at" to JsonPrimitive(42),
        ))

        val snapshot = projection.applyRoomState(freshSession)!!

        assertEquals("42", snapshot.getValue("updated_at").jsonPrimitive.content)
        assertEquals("2", snapshot.getValue("revision").jsonPrimitive.content)
    }

    @Test
    fun `tombstone prevents stale state from resurrecting a deleted session`() {
        val projection = MatrixNativeProjection()
        val current = sessionState("session-1", 2)
        val deleted = tombstone("session-1", 4)
        projection.applyRoomState(current)
        projection.applyRoomState(gatewayState(2))
        projection.applyRoomState(deleted)
        projection.applyRoomState(sessionState("session-1", 3, "Stale"))
        assertEquals(0, projection.snapshot()!!.sessions().size)
        assertEquals("deleted", projection.sessionLifecycleState("session-1"))
    }

    @Test
    fun `older tombstone cannot complete while newer active entity remains projected`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomState(gatewayState(2))
        projection.applyRoomState(sessionState("session-1", 5))
        projection.applyRoomState(tombstone("session-1", 4))

        assertEquals("active", projection.sessionLifecycleState("session-1"))
        assertEquals(1, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `new revision generation discards retired generation entities`() {
        val projection = MatrixNativeProjection()
        val old = sessionState("old", 2)
        projection.applyRoomState(old)
        projection.applyRoomState(gatewayState(2))
        val next = JsonObject(gatewayState(1) + mapOf(
            "revision_epoch" to JsonPrimitive("epoch-2"),
            "revision_epoch_generation" to JsonPrimitive(2),
        ))
        projection.applyRoomState(next)
        assertEquals(0, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `rejects incomplete or duplicate Gateway capabilities`() {
        val projection = MatrixNativeProjection()
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomState(JsonObject(gatewayState(1) + mapOf(
                "capabilities" to buildJsonObject { put("can_create_session", true) },
            )))
        }
        val duplicateModes = capabilities(permissionModeIds = listOf("default", "default"))
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomState(JsonObject(gatewayState(1) + mapOf(
                "capabilities" to duplicateModes,
            )))
        }
    }

    private fun gatewayState(version: Long) = buildJsonObject {
        put("version", 2); put("kind", "gateway_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("active_device_count", 1)
        put("command_sequences", JsonArray(listOf(buildJsonObject {
            put("device_id", "device-1")
            put("sequence_epoch", "certificate-1")
            put("sequence", 0)
        })))
        put("workspace", buildJsonObject {
            put("project", buildJsonObject {
                put("id", "project-1"); put("name", "malink"); put("cwd", "/repo")
            })
            put("provider", "codex"); put("permission_mode", "default")
        })
        put("capabilities", capabilities())
        put("session_directory", buildJsonObject {
            put("generation", version)
            put("state_version", version)
            put("slot", version % 3)
            put("page_count", 0)
            put("state_key_prefix", "malink.directory")
            put("digest", "RBNvo1WzZ4oRRq0W9-hknpT7T8If536DEMBg9hyq_4o")
        })
        put("updated_at", version)
    }

    private fun capabilities(
        permissionModeIds: List<String> = listOf("default"),
    ) = buildJsonObject {
        put("models", JsonArray(emptyList()))
        put("permission_modes", JsonArray(permissionModeIds.map { id ->
            buildJsonObject { put("id", id); put("name", "Default") }
        }))
        put("can_create_session", true)
        put("can_select_session", false)
        put("can_archive_session", true)
        put("can_delete_session", true)
        put("session_extensions", JsonArray(emptyList()))
    }

    private fun sessionState(id: String, version: Long, title: String = "Work") = buildJsonObject {
        put("version", 2); put("kind", "session_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("session_id", id); put("state", "active")
        put("session", buildJsonObject {
            put("session_id", id); put("title", title); put("updated_at", version)
            put("archived", false); put("status", "idle")
            put("project", buildJsonObject {
                put("id", "project-1"); put("name", "malink"); put("cwd", "/repo")
            })
            put("provider", "codex"); put("extensions", JsonArray(emptyList()))
        })
        put("updated_at", version)
    }

    private fun tombstone(id: String, version: Long) = buildJsonObject {
        put("version", 2); put("kind", "session_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("session_id", id); put("state", "deleted"); put("updated_at", version)
    }

    private fun kotlinx.serialization.json.JsonObjectBuilder.putRevision(revision: Long) {
        put("revision", revision); put("revision_epoch", "epoch-1")
        put("revision_epoch_generation", 1)
    }

    private fun JsonObject.sessions() = getValue("sessions").jsonArray.map { it.jsonObject }
    private fun JsonObject.text(key: String) = getValue(key).jsonPrimitive.content
}
